### DISCUSSION TITLE
[V5] Selectively disable CORS on sub-route, possible?

### DISCUSSION THREAD

**Discussion by aayush123** [adonisjs/core#2639] [Help] [answered]
I have two groups of routes:
1. User APIs under the group `api.example.com/api/v1`
2. Admin APIs under the group `api.example.com/api/v1/admin`

User APIs are accessed from a mobile app (React Native), so I have enabled cors and set allowed origins to '*'.
However, Admin APIs are accessed from an SPA (React) so I would like to restrict CORS to allow only SPA subdomain (`admin.example.com`)

I tried to disable CORS for the URL path starting with `api/v1/admin` but that breaks access as `admin.example.com` needs CORS to make requests to `api.example.com`.
Code snippet I used in `config/cors.ts`:
```typescript
...
  enabled: (request) => {
    if (Env.get('NODE_ENV') !== 'development' && request.url().startsWith('/api/v1/admin')) {
      return false;
    }
    return true;
  },
  origin: '*',
...
```

Is there any way to do it using the cors config file? Or even through some middleware?
Also, is CORS needed for APIs consumed in mobile apps?

**Comment by thetutlage** [comment, marked as answer]
You can use the `request.hostname` to check for the hostname. I will also simplify the handler as follows to allow `admin.example.com` for all endpoints

```ts
  enabled: (request) => {
    if (Application.inDev) {
      return true
    }

    return request.hostname() === 'admin.example.com'
  },
```

> Also, I don't think the mobile native apps has anything to do with CORS at all. But still double check, as I don't create mobile apps

**Reply by aayush123** [reply]
> Also, I don't think the mobile native apps has anything to do with CORS at all. But still double check, as I don't create mobile apps

Yeah, it doesn't matter on mobile apps. This also makes sense cause CORS is enforced by a client. Mobile apps don't usually have anything to enforce CORS.

@thetutlage  Also, I had to change the condition in `enabled` function to:
```typescript
  enabled: (request) => {
    if (Application.inDev) {
      return true;
    }
    return request.url().startsWith('/api/v1/admin');  // 👈 
  },
  origin: [Application.inDev ? 'http://localhost' : 'https://admin.qwiggo.com'],
```
As apparently request.hostname() gives the hostname in the API called, ie. `api.example.com`, which does not give info about the origin host, but rather the _destination host_.
I added the conditional host in `origin` key just to be extra safe in prod 😅

**Comment by thetutlage** [comment]
Ohh yeah. My bad. hostname returns the current server hostname and not the client

**Reply by aayush123** [reply]
No worries mate. Thanks for the framework though. Absolutely love it! :)
